### PR TITLE
Feature/media preview image

### PR DESCRIPTION
**Pull request by janmassive**

### CODE REVIEW

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 1]
```diff
@@ -211,6 +211,99 @@ public function deleteAction($id)
     }
 
     /**
+     * Creates a new preview image and saves it to the provided media.
+     *
+     * @Post("media/preview/{id}")
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: That's not REST conform, should be solved as child resource, as in e.g. https://github.com/sulu-io/sulu/blob/develop/src/Sulu/Bundle/ContactBundle/Controller/ContactMediaController.php

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 2]
```diff
@@ -211,6 +211,99 @@ public function deleteAction($id)
...
+            $mediaManager = $this->getMediaManager();
+            $systemCollectionManager = $this->get('sulu_media.system_collections.manager');
+
+            $locale = $this->getLocale($request);
+
+            $media = $mediaManager->getById($id, $locale);
+            $mediaEntity = $media->getEntity();
+
+            $data = $this->getData($request, false);
+            unset($data['id']);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: Why is this necessary?

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 3]
```diff
@@ -211,6 +211,99 @@ public function deleteAction($id)
...
+            $uploadedFile = $this->getUploadedFile($request, 'previewImg');
+            $previewImg = $mediaManager->save($uploadedFile, $data, $this->getUser()->getId());
+
+            $mediaEntity->setPreviewImage($previewImg->getEntity());
+            $mediaManager->saveEntity($mediaEntity);
+
+            // Add preview thumbnails
+            $media = $mediaManager->addFormatsAndUrl($media);
+
+            $view = $this->view($media, 200);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: When this is an own Controller, it should return the preview image here.

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 4]
```diff
@@ -211,6 +211,99 @@ public function deleteAction($id)
...
+            $view = $this->view($e->toArray(), 400);
+        }
+
+        return $this->handleView($view);
+    }
+
+    /**
+     * Removes current preview image and sets default video thumbnail.
+     *
+     * @Get("media/reset-preview/{id}")
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: A get should never delete something... This should also go in the child resource I've mentioned in the previous method. Then it can be a delete action.

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 5]
```diff
@@ -211,6 +211,99 @@ public function deleteAction($id)
...
+
+            $mediaEntity->setPreviewImage(null);
+
+            $mediaManager->saveEntity($mediaEntity);
+            $mediaManager->delete($oldPreviewImageId);
+
+            // Reload thumbnails
+            $media = $mediaManager->addFormatsAndUrl($media);
+
+            $view = $this->view($media, 200);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: Since there are no preview images after this action, it should probably return a 204 with an empty response.

File: src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php [hunk 6]
```diff
@@ -447,6 +447,15 @@ public function save($uploadedFile, $data, $userId)
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function saveEntity(\Sulu\Bundle\MediaBundle\Entity\Media $media)
+    {
+        $this->em->persist($media);
+        $this->em->flush();
+    }
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php]: I would not introduce this function... Also, where you are using it, you don't need the call to persist, so I would just call flush there. (Although I am not sure if that you need something like this is the best idea either...) Isn't there a possibility that you can use one of the existing methods.

File: src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php [hunk 7]
```diff
@@ -906,6 +915,42 @@ public function addFormatsAndUrl(Media $media)
...
+        // Get preview image and overwrite thumbnails if set
+        /** @var \Sulu\Bundle\MediaBundle\Entity\Media $previewImage */
+        $previewImage = $media->getEntity()->getPreviewImage();
+
+        if ($previewImage !== null) {
+            /** @var FileVersion $latestVersion */
+            $latestVersion = null;
+
+            /** @var File $file */
+            foreach ($previewImage->getFiles() as $file) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php]: There should be a version variable on one of the entities containing the latest version, so you don't have to loop over all the versions.

File: src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php [hunk 8]
```diff
@@ -906,6 +915,42 @@ public function addFormatsAndUrl(Media $media)
...
+                        $highestVersion = $version->getVersion();
+                    }
+                }
+
+                // currently only one file per media exists
+                break;
+            }
+
+            if ($latestVersion !== null) {
+                $media->setFormats(
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php]: I can't imagine that we really want to override the existing formats here. How can you then access the video itself?

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js [hunk 9]
```diff
@@ -179,6 +188,7 @@ define([
          * @param media {Object} the id of the media to edit
          */
         editSingleMedia: function(media) {
+            var $preview;
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js]: There should be only one var statement in each function, so move this one together with `$info`

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js [hunk 10]
```diff
@@ -344,6 +418,35 @@ define([
...
+                        el: constants.previewDropzoneSelector,
+                        maxFilesize: config.get('sulu-media').maxFilesize,
+                        url: '/admin/api/media/preview/' + this.media.id,
+                        method: 'POST',
+                        paramName: 'previewImg',
+                        showOverlay: false,
+                        skin: 'overlay',
+                        titleKey: '',
+                        descriptionKey: 'sulu.media.upload-new-preview',
+                        instanceName: 'preview-img',
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js]: We try to avoid abbreviations like `img` (same for the `paramName` a few lines earlier)

File: src/Sulu/Bundle/MediaBundle/Resources/public/scss/modules/media-edit-overlay.scss [hunk 11]
```diff
@@ -1,46 +1,69 @@
...
+    #file-version-change {
+        width: 100%;
+        height: 100%;
+
+        .dropzone-container {
+            width: 150px;
+            height: 150px;
+        }
+    }
+
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/scss/modules/media-edit-overlay.scss]: no new line here

**Comment by danrot** [comment]
Did you have a look at the design with @berndhep ?

**Comment by janmassive** [comment]
Ready for review

File: src/Sulu/Bundle/MediaBundle/Controller/AbstractMediaController.php [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+namespace Sulu\Bundle\MediaBundle\Controller;
+
+use Sulu\Bundle\MediaBundle\Media\Manager\MediaManagerInterface;
+use Sulu\Component\Rest\RestController;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\HttpFoundation\Request;
+
+class AbstractMediaController extends RestController
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/AbstractMediaController.php]: Missing PHPDoc

File: src/Sulu/Bundle/MediaBundle/Controller/AbstractMediaController.php [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+     * @return string
+     */
+    protected function getTitleFromUpload($request)
+    {
+        $title = null;
+
+        $uploadedFile = $this->getUploadedFile($request, 'fileVersion');
+
+        if ($uploadedFile) {
+            $title = $part = implode('.', explode('.', $uploadedFile->getClientOriginalName(), -1));
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/AbstractMediaController.php]: What's the `$part` for?

File: src/Sulu/Bundle/MediaBundle/Controller/MediaPreviewController.php [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+ * Makes medias preview images available through a REST API.
+ *
+ * @RouteResource("Preview")
+ */
+class MediaPreviewController extends AbstractMediaController implements ClassResourceInterface
+{
+    /**
+     * Creates a new preview image and saves it to the provided media.
+     *
+     * @Post("media/{id}/preview")
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaPreviewController.php]: These shouldn't be necessary anymore at all

File: src/Sulu/Bundle/MediaBundle/Controller/MediaPreviewController.php [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+            $locale = $this->getLocale($request);
+
+            $media = $mediaManager->getById($id, $locale);
+            /** @var Media $mediaEntity */
+            $mediaEntity = $media->getEntity();
+
+            $data = $this->getData($request, false);
+
+            // Unset id to not overwrite original file
+            unset($data['id']);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaPreviewController.php]: Still needed?

File: src/Sulu/Bundle/MediaBundle/Controller/MediaPreviewController.php [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+            if ($mediaEntity->getPreviewImage() !== null) {
+                $data['id'] = $mediaEntity->getPreviewImage()->getId();
+            }
+            $data['collection'] = $systemCollectionManager->getSystemCollection('sulu_media.preview_img');
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaPreviewController.php]: There's still the `img` abbreviation :wink:

**Comment by danrot** [comment]
@janmassive Some feedback from @berndhep and @chirimoya 

* The translation for the new tab should be "Vorschaubild" resp. "Preview image".
* The history tab should be the last one.
* In the preview tab the user should only be able to upload images (no documents, videos, ...)
* The preview image should be displayed for every media type except images.

**Comment by danrot** [comment]
Rebased and merged
